File: backend/src/main/java/com/codeloom/backend/controller/TopicController.java
package com.codeloom.backend.controller;

import com.codeloom.backend.dto.CreateTopicRequest;
import com.codeloom.backend.dto.UpdateTopicRequest;
import com.codeloom.backend.model.Topic;
import com.codeloom.backend.service.TopicService;
import jakarta.validation.Valid;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/v1/topics")
@RequiredArgsConstructor
public class TopicController {
    private final TopicService service;

    @GetMapping
    public Iterable<Topic> findAll() {
        return service.findAll();
    }

    @GetMapping("/{id}")
    public Topic findAll(@PathVariable UUID id) {
        return service.findById(id);
    }

    @PostMapping
    public Topic create(@Valid @RequestBody CreateTopicRequest request) {
        return service.create(request);
    }

    @PutMapping("/{id}")
    public Topic update(@PathVariable UUID id, @Valid @RequestBody UpdateTopicRequest request) {
        return service.update(id, request);
    }

    @DeleteMapping("/{id}")
    public void delete(@PathVariable UUID id) {
        service.delete(id);
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/controller/TopicController.kt
package com.codeloom.backend.controller

import com.codeloom.backend.dto.CreateTopicRequest
import com.codeloom.backend.model.Topic
import com.codeloom.backend.service.TopicService
import jakarta.validation.Valid
import org.springframework.web.bind.annotation.*
import tools.jackson.databind.JsonNode
import java.util.*

@RestController
@RequestMapping("/v1/topics")
class TopicController(private val topicService: TopicService) {
    @GetMapping
    fun findAll(): Iterable<Topic> = topicService.findAll()

    @GetMapping("/{id}")
    fun findAll(@PathVariable id: UUID): Topic = topicService.findById(id)

    @PostMapping
    fun create(@Valid @RequestBody request: CreateTopicRequest): Topic = topicService.create(request)

    @PatchMapping("/{id}")
    fun patch(@PathVariable id: UUID, @RequestBody patchNode: JsonNode): Topic = topicService.patch(id, patchNode)

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: UUID) = topicService.delete(id)
}
